README.md
# osx_dotfiles


ranger/history
open_with vim
Q


swm/swc/swc.c
#include <stdio.h>
#include <xcb/xcb.h>
#include <err.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/shm.h>

#define BORDERWIDTH 4

static xcb_connection_t *conn;
static xcb_screen_t *scr;

static void cleanup (void);
static void killwin (xcb_window_t);
static void center_pointer (xcb_window_t);
static void nextwin (xcb_window_t);
static void resize (xcb_window_t, int, int);
static void move (xcb_window_t, int, int);

static void
cleanup (void) {
	if (conn)
		xcb_disconnect(conn);
}

static void
killwin (xcb_window_t win) {
	if (!win || win == scr->root)
		return;

	xcb_kill_client(conn, win);
	nextwin(win);
}

static void
center_pointer (xcb_window_t win) {
	uint32_t values[1];
	xcb_get_geometry_reply_t *geom;
	geom = xcb_get_geometry_reply(conn, xcb_get_geometry(conn, win), NULL);

	if (!geom)
		errx(1, "center_pointer: missing geometry!");

	xcb_warp_pointer(conn, XCB_NONE, win, 0, 0, 0, 0,
			(geom->width  + (BORDERWIDTH * 2)) / 2,
			(geom->height + (BORDERWIDTH * 2)) / 2);

	values[0] = XCB_STACK_MODE_ABOVE;
	xcb_configure_window(conn, win, XCB_CONFIG_WINDOW_STACK_MODE, values);
}

static void
nextwin (xcb_window_t win) {
	xcb_query_tree_reply_t *r;
	xcb_window_t *c,t = 0;
	xcb_get_window_attributes_cookie_t ac;
	xcb_get_window_attributes_reply_t *ar;

	r = xcb_query_tree_reply(conn, xcb_query_tree(conn, scr->root), 0);
	if (!r || !r->children_len)
		return;
	c = xcb_query_tree_children(r);
	for (unsigned int i=0; i < r->children_len; i++) {
		ac = xcb_get_window_attributes(conn, c[i]);
		ar = xcb_get_window_attributes_reply(conn, ac, NULL);

		if (ar && ar->map_state == XCB_MAP_STATE_VIEWABLE) {
			if (!(ar->override_redirect || c[i] == win)) {
				t = c[i];
			}
			break;
		}
	}

	if (t)
		center_pointer(t);

	free(r);
}

static void
resize (xcb_window_t win, int x, int y) {
	uint32_t values[2];
	int real;
	xcb_get_geometry_reply_t *geom;

	if (!win || win == scr->root)
		return;
	geom = xcb_get_geometry_reply(conn, xcb_get_geometry(conn, win), NULL);

	if (!geom)
		return;

	values[0] = x ? geom->width + x : geom->width;
	values[1] = y ? geom->height + y : geom->height;

	if (x)
	{
		real = geom->width + (BORDERWIDTH * 2);
		if (geom->x + real + x > scr->width_in_pixels)
			values[0] = scr->width_in_pixels - geom->x -
				(BORDERWIDTH * 2);
	}

	if (y)
	{
		real = geom->height + (BORDERWIDTH *2);
		if (geom->y + real + y > scr->height_in_pixels)
			values[1] = scr->height_in_pixels - geom->y -
				(BORDERWIDTH * 2);
	}

	xcb_configure_window(conn, win, XCB_CONFIG_WINDOW_WIDTH
			| XCB_CONFIG_WINDOW_HEIGHT, values);

	center_pointer(win);

	free(geom);
}

static void
move (xcb_window_t win, int x, int y) {
	uint32_t values[2];
	int real;
	xcb_get_geometry_reply_t *geom;

	if (!win || win == scr->root)
		return;

	geom = xcb_get_geometry_reply(conn, xcb_get_geometry(conn, win), NULL);
	if (!geom)
		return;

	values[0] = x ? geom->x + x : geom->x;
	values[1] = y ? geom->y + y : geom->y;

	if (x)
	{
		real = geom->width + (BORDERWIDTH * 2);
		if (geom->x + x < 1)
			values[0] = 0;
		if (geom->x + x > scr->width_in_pixels - real)
			values[0] = scr->width_in_pixels - real;
	}

	if (y)
	{
		real = geom->height + (BORDERWIDTH * 2);
		if (geom->y + y < 1)
			values[1] = 0;
		if (geom->y + y > scr->height_in_pixels - real)
			values[1] = scr->height_in_pixels - real;
	}

	xcb_configure_window(conn, win, XCB_CONFIG_WINDOW_X
			| XCB_CONFIG_WINDOW_Y, values);


	center_pointer(win);
	free(geom);
}

int main (int argc, char **argv) {
	atexit(cleanup);
	xcb_window_t *focuswin;
	xcb_window_t win;
	if (xcb_connection_has_error(conn = xcb_connect(NULL, NULL)))
		errx(1, "error connecting to xcb");

	scr = xcb_setup_roots_iterator(xcb_get_setup(conn)).data;

	key_t key = 0x0DEADCA7;
	int shmid;

	if ((shmid = shmget(key, sizeof(xcb_window_t), 0666)) < 0)
		warnx("shmget error");

	if ((focuswin = shmat(shmid, NULL, 0)) == (xcb_window_t *) -1)
		warnx("shmat error");

	win  = *focuswin;
	if (!win)
		errx(1, "no window");

	if (argc >= 2 && argv[1][0])
		switch (argv[1][1])
		{
			case 'm':
				if (argc != 4)
					return 1;
				
				switch (argv[2][0])
				{
					case 'x':
					case 'X':
						move(win, atoi(argv[3]), 0);
						break;
					case 'y':
					case 'Y':
						move(win, 0, atoi(argv[3]));
						break;
				}
			break;

			case 'r':
				if (argc != 4)
					return 1;

				switch (argv[2][0])
				{
					case 'x':
					case 'X':
						resize(win, atoi(argv[3]), 0);
						break;
					case 'y':
					case 'Y':
						resize(win, 0, atoi(argv[3]));
						break;
				}
			break;

			case 'f':
				if (argc != 3)
					return 1;

				switch (argv[2][0])
				{
					case 'n':
					case 'N':
						nextwin(win);
						break;
				}
			break;

			case 'k':
				killwin(win);
			break;
		}

	xcb_flush(conn);

	return 0;
}


homepage.py/homepage.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

########## begin config ##########

output = 'homepage.html'      # output file
title = 'Homepage'          # page title
font = 'fixed'  # font
separator = '    »  '   	# separator between group title and links
#colors = (
#    '#181512',  # background
#    '#bea492',  # links
#    '#646a6d',  # group title 
#    '#8c644c',  # separator
#)
colors = (
    '#181512',  # background
    '#bea492',  # links
    '#9a875f',  # group title 
    '#504339',  # separator
)

# add your links here
links = {
		'search': [
			['google', 'https://www.google.com/'],
			['duckduckgo', 'http://duckduckgo.com/'],
			['wiki', 'http://en.wikipedia.org/wiki/Main_Page'],
			['images', 'https://images.google.com/'],
		],
		'social': [
			['twitter', 'http://www.twitter.com'],
			['gmail', 'http://mail.google.com'],
			['youtube', 'http://www.youtube.com/'],
			['tumblr', 'http://www.tumblr.com/'],
			['github', 'http://www.github.com/'],
			['last.fm', 'http://www.last.fm/'],
			['deviantart', 'http://www.deviantart.com/'],
		],
		'news': [
			['reddit', 'http://www.reddit.com'],
			['hackernews', 'http://news.ycombinator.com'],
			['arch', 'http://bbs.archlinux.org'],
			['crunch', 'http://www.crunchbanglinux.org/forums'],
			['omgubuntu', 'http://www.omgubuntu.co.uk'],
		],
		'private': [
                        ['vps', 'http://b4dtr1p.tk'],
                        ['git', 'http://git.b4dtr1p.tk'],
		],
}

########## end config ##########

import os

css = '''body {
  background-color: %s;
  font-family: %s;
  font-size: %s;
  font-weight: normal;
  margin-left: 7%%;
}
a:link,a:visited,a:active {
  text-decoration: none;
  color: %s;
  font-weight: normal;
}
a:hover {
  text-decoration: underline;
  color: %s;
  font-weight: normal;
}
table {
  border-spacing: 8px;
}
td:first-child {
  font-weight: bold;
  color: %s
}
td:nth-child(2) {
  font-weight: normal;
  color: %s;
}''' % (colors[0], font[0], font[1], colors[1], colors[1], colors[2], colors[3])

links_html = ''
for group in links:
    links_html += '<tr><td align="right">%s</td><td>%s</td><td>' % (group, separator)
    for site in links[group]:
        links_html += '<a href="%s">%s</a> ' % (site[1], site[0])
    links_html += '</td></tr>'

html = '''<html>
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
  <title>%s</title>
  <link rel="stylesheet" type="text/css" href="style.css" />
</head>
<body>
  <table valign="middle" border="0" width="100%%" height="100%%"><tr><td><table>
    %s
  </table></td></tr></table>
</body>
</html>''' % (title, links_html)

with open(output, 'w') as file:
    file.write(html)
with open('style.css', 'w') as file:
    file.write(css)

os.system('tidy -utf8 -i -m -q -asxhtml %s' % output)


swm/swm.c
/**
*      Copyright (c) 2014, Broseph <dcat (at) iotek (dot) org>
*
*      Permission to use, copy, modify, and/or distribute this software for any
*      purpose with or without fee is hereby granted, provided that the above
*      copyright notice and this permission notice appear in all copies.
*
*      THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
*      WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
*      MERCHANTABILITY AND FITNESS IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
*      ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
*      WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
*      ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
*      OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
**/

#define _XOPEN_SOURCE	/* for shm */
#define CLEANMASK(m)	((m & ~0x80))

/* sane modifier names */
#define SUPER		XCB_MOD_MASK_4
#define ALT		XCB_MOD_MASK_1
#define CTRL		XCB_MOD_MASK_CONTROL
#define SHIFT		XCB_MOD_MASK_SHIFT

#include <err.h>
#include <stdio.h>
#include <signal.h>
#include <stdlib.h>
#include <unistd.h>
#include <xcb/xcb.h>
#include <sys/shm.h>

enum { INACTIVE, ACTIVE, };

/* global variables */
static xcb_connection_t		*conn;
static xcb_screen_t		*scr;
static xcb_window_t		*focuswin;

/* proto */
static void cleanup		(void);
static void swm_init		(void);
static void setup_win		(xcb_window_t);
static void focus		(xcb_window_t, int);

#include "config.h"

static void
cleanup (void) {
	/* graceful exit */
	if (conn)
		xcb_disconnect(conn);
}

static void
swm_init (void) {
	uint32_t values[2];
	key_t key = 0x0DEADCA7;
	int shmid;

	if (xcb_connection_has_error(conn = xcb_connect(NULL, NULL)))
		errx(1, "error connecting to xcb");

	scr = xcb_setup_roots_iterator(xcb_get_setup(conn)).data;

	if ((shmid = shmget(key, sizeof(xcb_window_t), IPC_CREAT | 0666)) < 0)
		errx(1, "shmget error");

	if ((focuswin = shmat(shmid, NULL, 0)) == (xcb_window_t *) -1)
		errx(1, "shmat error");

	*focuswin = scr->root;

#ifdef ENABLE_MOUSE
	xcb_grab_button(conn, 0, scr->root, XCB_EVENT_MASK_BUTTON_PRESS |
			XCB_EVENT_MASK_BUTTON_RELEASE, XCB_GRAB_MODE_ASYNC,
			XCB_GRAB_MODE_ASYNC, scr->root, XCB_NONE, 1, MOD);

	xcb_grab_button(conn, 0, scr->root, XCB_EVENT_MASK_BUTTON_PRESS |
			XCB_EVENT_MASK_BUTTON_RELEASE, XCB_GRAB_MODE_ASYNC,
			XCB_GRAB_MODE_ASYNC, scr->root, XCB_NONE, 3, MOD);
#endif

	values[0] = XCB_EVENT_MASK_SUBSTRUCTURE_NOTIFY;
	xcb_change_window_attributes_checked(conn, scr->root, XCB_CW_EVENT_MASK,
			values);

	xcb_flush(conn);
}

static void
focus (xcb_window_t win, int mode) {
	uint32_t values[1];
	short w, h, b, o;

	if (!win)
		return;

	xcb_get_geometry_reply_t *geom = xcb_get_geometry_reply(conn,
			xcb_get_geometry(conn, win), NULL);

	if (!geom)
		return;

	w = (short)geom->width;
	h = (short)geom->height;
	b = (unsigned short)BORDERWIDTH;
	o = (unsigned short)OUTER;

	xcb_rectangle_t inner[] = {
		/* you're not supposed to understand this. */
		{     w,0,b-o     ,h+b-   o      },
		{     w+b   +o,  0,   b  -o,     h+         b  -  o},
		{     0,h   ,w+b  -o,b-   o      },
		{     0,h   +b+      o,   w+     b-         o, b -o},
		{     w+b+o,b        +h    +o,b,b}
	};

	xcb_rectangle_t outer[] = {
		{w + b - o, 0, o, h + b * 2},
		{w + b,     0, o, h + b * 2},
		{0, h + b - o, w + b * 2, o},
		{0, h + b,     w + b * 2, o},
		{1, 1, 1, 1}
	};

	xcb_pixmap_t pmap = xcb_generate_id(conn);
	xcb_create_pixmap(conn, scr->root_depth, pmap, win, geom->width
			+ (BORDERWIDTH * 2), geom->height + (BORDERWIDTH * 2));
	xcb_gcontext_t gc = xcb_generate_id(conn);
	xcb_create_gc(conn, gc, pmap, 0, NULL);

	values[0] = OUTER_COLOR;
	xcb_change_gc(conn, gc, XCB_GC_FOREGROUND, values);
	xcb_poly_fill_rectangle(conn, pmap, gc, 5, outer);

	values[0] = mode ? FOCUSCOL : UNFOCUSCOL;
	xcb_change_gc(conn, gc, XCB_GC_FOREGROUND, values);
	xcb_poly_fill_rectangle(conn, pmap, gc, 5, inner);

	values[0] = pmap;
	xcb_change_window_attributes(conn, win, XCB_CW_BORDER_PIXMAP, values);

	if (mode) {
		xcb_set_input_focus(conn, XCB_INPUT_FOCUS_POINTER_ROOT,
				win, XCB_CURRENT_TIME);
		if (win != *focuswin) {
			focus(*focuswin, INACTIVE);
			*focuswin = win;
		}
	}

	xcb_free_pixmap(conn, pmap);
	xcb_free_gc(conn, gc);
}

static void
setup_win (xcb_window_t win) {
	uint32_t values[2];

	values[0] = XCB_EVENT_MASK_ENTER_WINDOW;
	values[1] = XCB_EVENT_MASK_SUBSTRUCTURE_NOTIFY;
	xcb_change_window_attributes(conn, win, XCB_CW_EVENT_MASK, values);

	/* border width */
	values[0] = BORDERWIDTH;
	values[1] = XCB_NONE;
	xcb_configure_window(conn, win, XCB_CONFIG_WINDOW_BORDER_WIDTH, values);
}

static void
events_loop (void) {
	xcb_generic_event_t *ev;
#ifdef ENABLE_MOUSE
	uint32_t values[3];
	xcb_get_geometry_reply_t *geom;
	xcb_window_t win = 0;
#endif

	/* loop */
	for (;;) {
		ev = xcb_wait_for_event(conn);

		if (!ev)
			errx(1, "xcb connection broken");

		switch (CLEANMASK(ev->response_type)) {

		case XCB_CREATE_NOTIFY: {
			xcb_create_notify_event_t *e;
			e = (xcb_create_notify_event_t *)ev;

			if (!e->override_redirect) {
				setup_win(e->window);
				focus(e->window, ACTIVE);
			}
		} break;

		case XCB_DESTROY_NOTIFY: {
			xcb_destroy_notify_event_t *e;
			e = (xcb_destroy_notify_event_t *)ev;

			xcb_kill_client(conn, e->window);
		} break;

		case XCB_ENTER_NOTIFY: {
			xcb_enter_notify_event_t *e;
			e = (xcb_enter_notify_event_t *)ev;

			focus(e->event, ACTIVE);
		} break;

		case XCB_MAP_NOTIFY: {
			xcb_map_notify_event_t *e;
			e = (xcb_map_notify_event_t *)ev;

			if (!e->override_redirect) {
				xcb_map_window(conn, e->window);
				xcb_set_input_focus(conn,
					XCB_INPUT_FOCUS_POINTER_ROOT,
					e->window, XCB_CURRENT_TIME);
			}
		} break;

#ifdef ENABLE_MOUSE
		case XCB_BUTTON_PRESS: {
			xcb_button_press_event_t *e;
			e = ( xcb_button_press_event_t *)ev;
			win = e->child;

			if (!win || win == scr->root)
				break;

			values[0] = XCB_STACK_MODE_ABOVE;
			xcb_configure_window(conn, win,
					XCB_CONFIG_WINDOW_STACK_MODE, values);
			geom = xcb_get_geometry_reply(conn,
					xcb_get_geometry(conn, win), NULL);
			if (1 == e->detail) {
				values[2] = 1;
				xcb_warp_pointer(conn, XCB_NONE, win, 0, 0, 0,
					0, geom->width/2, geom->height/2);
			} else {
				values[2] = 3;
				xcb_warp_pointer(conn, XCB_NONE, win, 0, 0, 0,
						0, geom->width, geom->height);
			}
			xcb_grab_pointer(conn, 0, scr->root,
				XCB_EVENT_MASK_BUTTON_RELEASE
				| XCB_EVENT_MASK_BUTTON_MOTION
				| XCB_EVENT_MASK_POINTER_MOTION_HINT,
				XCB_GRAB_MODE_ASYNC, XCB_GRAB_MODE_ASYNC,
				scr->root, XCB_NONE, XCB_CURRENT_TIME);
			xcb_flush(conn);
		} break;

		case XCB_MOTION_NOTIFY: {
			xcb_query_pointer_reply_t *pointer;
			pointer = xcb_query_pointer_reply(conn,
					xcb_query_pointer(conn, scr->root), 0);
			if (values[2] == 1) {
				geom = xcb_get_geometry_reply(conn,
					xcb_get_geometry(conn, win), NULL);
				if (!geom) {
					break;
				}

				values[0] = (pointer->root_x + geom->width / 2
					> scr->width_in_pixels
					- (BORDERWIDTH*2))
					? scr->width_in_pixels - geom->width
					- (BORDERWIDTH*2)
					: pointer->root_x - geom->width / 2;
				values[1] = (pointer->root_y + geom->height / 2
					> scr->height_in_pixels
					- (BORDERWIDTH*2))
					? (scr->height_in_pixels - geom->height
					- (BORDERWIDTH*2))
					: pointer->root_y - geom->height / 2;

				if (pointer->root_x < geom->width/2)
					values[0] = 0;
				if (pointer->root_y < geom->height/2)
					values[1] = 0;

				xcb_configure_window(conn, win,
					XCB_CONFIG_WINDOW_X
					| XCB_CONFIG_WINDOW_Y, values);
				xcb_flush(conn);
			} else if (values[2] == 3) {
				geom = xcb_get_geometry_reply(conn,
					xcb_get_geometry(conn, win), NULL);
				values[0] = pointer->root_x - geom->x;
				values[1] = pointer->root_y - geom->y;
				xcb_configure_window(conn, win,
					XCB_CONFIG_WINDOW_WIDTH
					| XCB_CONFIG_WINDOW_HEIGHT, values);
				xcb_flush(conn);
			}
		} break;

		case XCB_BUTTON_RELEASE:
			focus(win, ACTIVE);
			xcb_ungrab_pointer(conn, XCB_CURRENT_TIME);
			break;
#endif

		case XCB_CONFIGURE_NOTIFY: {
			xcb_configure_notify_event_t *e;
			e = (xcb_configure_notify_event_t *)ev;

			focus(*focuswin, ACTIVE);
		} break;

		}

		xcb_flush(conn);
		free(ev);
	}
}

int
main (void) {
	/* graceful exit */
	atexit(cleanup);

	swm_init();
	events_loop();

	return EXIT_FAILURE;
}

/* vim: set noet sw=8 sts=8: */
